src/shims/io.rs

```diff
@@ -0,0 +1,170 @@
+            flags: Default::default(),
+            // 0, 1 and 2 are reserved for stdin, stdout and stderr
+            low: 3,
+        }
+    }
+}
+
+impl<'mir, 'tcx> EvalContextExt<'mir, 'tcx> for crate::MiriEvalContext<'mir, 'tcx> {}
+pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx> {
+    fn open(
+        &mut self,
+        path_op: OpTy<'tcx, Tag>,
+        flag_op: OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        if !this.machine.communicate {
+            throw_unsup_format!("`open` not available when isolation is enabled")
+        }
+
+        let flag = this.read_scalar(flag_op)?.to_i32()?;
+
+        if flag != this.eval_libc_i32("O_RDONLY")? && flag != this.eval_libc_i32("O_CLOEXEC")? {
+            throw_unsup_format!("Unsupported flag {:#x}", flag);
+        }
+
+        let path_bytes = this
+            .memory()
+            .read_c_str(this.read_scalar(path_op)?.not_undef()?)?;
+        let path = std::str::from_utf8(path_bytes)
+            .map_err(|_| err_unsup_format!("{:?} is not a valid utf-8 string", path_bytes))?;
+
+        match File::open(path) {
+            Ok(file) => {
+                let mut fh = &mut this.machine.file_handler;
+                fh.low += 1;
+                fh.files.insert(fh.low, file);
+                fh.flags.insert(fh.low, flag);
+                Ok(fh.low)
+            }
+
+            Err(e) => {
+                this.machine.last_error = e.raw_os_error().unwrap() as u32;
+                Ok(-1)
+            }
+        }
+    }
+
+    fn fcntl(
+        &mut self,
+        fd_op: OpTy<'tcx, Tag>,
+        cmd_op: OpTy<'tcx, Tag>,
+        arg_op: Option<OpTy<'tcx, Tag>>,
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        if !this.machine.communicate {
+            throw_unsup_format!("`open` not available when isolation is enabled")
+        }
+
+        let fd = this.read_scalar(fd_op)?.to_i32()?;
+        let cmd = this.read_scalar(cmd_op)?.to_i32()?;
+
+        if cmd == this.eval_libc_i32("F_SETFD")? {
+            // This does not affect the file itself. Certain flags might require changing the file
+            // or the way it is accessed somehow.
+            let flag = this.read_scalar(arg_op.unwrap())?.to_i32()?;
+            // The only usage of this in stdlib at the moment is to enable the `FD_CLOEXEC` flag.
+            let fd_cloexec = this.eval_libc_i32("FD_CLOEXEC")?;
+            if let Some(old_flag) = this.machine.file_handler.flags.get_mut(&fd) {
+                if flag ^ *old_flag == fd_cloexec {
+                    *old_flag = flag;
+                } else {
+                    throw_unsup_format!("Unsupported arg {:#x} for `F_SETFD`", flag);
+                }
+            }
+            Ok(0)
+        } else if cmd == this.eval_libc_i32("F_GETFD")? {
+            if let Some(flag) = this.machine.file_handler.flags.get(&fd) {
```

Comment:
This get-or-EBADF business should get its own method (after the `flags`/`files` merger this will be usable everywhere I think)

Reply:
such a method could even take a `OpTy<'tcx, Tag>` for the fd directly

Reply:
You mean as a method for the `FileHandler`, right?

Reply:
> such a method could even take a `OpTy<'tcx, Tag>` for the fd directly

I think that just taking the `i32` is better because taking the `OpTy` would mean evaluating it several times.

Reply:
oh, yea in that case just use the i32

Reply:
> You mean as a method for the FileHandler, right?

yes

Reply:
I think this method and the `consume_result` one shouldn't be public at all, what do you think?